### PR TITLE
feat: pack_context tool — token-budget-aware memory retrieval

### LINKED ISSUE #452: Add pack_context tool — token-budget-aware memory retrieval
## Problem

An agent asking "what do I know about X" often wants **as much relevant context as will fit in its remaining token budget**, not just top-K results or all memories tagged with X. `search_memories` returns a fixed number; `list_memories` returns everything under a tag; neither considers the agent's available context window.

The result is that agents either:
- Over-request and blow the context window, or
- Under-request and miss relevant memory they had budget for

## Proposed tool

```python
@mcp.tool
async def pack_context(
    topic: str,
    budget_tokens: int = 2000,
    ordering: Literal["relevance", "recency", "relevance+recency"] = "relevance+recency",
) -> str:
    """
    Retrieve as many relevant memories as fit within `budget_tokens`,
    ordered by `ordering`, and return them concatenated as a single
    context block ready to drop into the agent's prompt.
    """
```

### Behaviour

1. Run semantic search for `topic`
2. Score candidates by `ordering` (relevance, recency, or a weighted blend)
3. Estimate tokens for each candidate (tiktoken or heuristic — 1 token ≈ 4 chars)
4. Greedily pack until `budget_tokens` is reached
5. Return the packed context with a header noting how many memories were included and the effective budget used

### Output shape

```
## Context for 'X' (12 memories, ~1847 tokens)

- **preferences/python-formatter**: ruff with line length 100
- **decisions/auth**: JWT tokens stored in DynamoDB with 24h TTL
- ...
```

Plain text works — MCP tool responses are text-by-default and agents can paste directly into their working context.

## Why not do this client-side

- The agent doesn't know how expensive each memory is without fetching them all first — a single packing tool avoids the N+1 round-trip
- Token-aware truncation logic belongs in the server where memory metadata is cheap to inspect
- Centralising the logic means every client benefits equally

## Scope

- [ ] New `pack_context` tool in `src/hive/server.py`
- [ ] Token estimation helper (tiktoken or simple heuristic)
- [ ] Pagination by budget, not count
- [ ] Unit tests covering budget edge cases (budget < smallest memory, budget >> total corpus)
- [ ] Docs: `docs-site/tools/pack-context.md`

## Acceptance criteria

- Agent can call `pack_context(topic, budget_tokens=N)` and receive a response ≤ N tokens
- Ordering modes produce stable, deterministic results
- 100% coverage

## Why this matters

- Real agent pain point — manual context packing is error-prone
- Differentiator: no other MCP memory server (that I know of) offers token-budgeted retrieval
- Pairs naturally with #374 / sampling-based synthesis

### PR DESCRIPTION
**Pull request by warlordofmars**
Closes #452

## Summary

Ships a new `pack_context` MCP tool that retrieves **as many relevant
memories as fit in a caller-supplied token budget**, returned as a
markdown block ready to drop into the agent's prompt. Unlike
`search_memories` (fixed top-K) and `list_memories` (full tag slice),
this is the first retrieval surface that understands the agent's
remaining context window.

| Arg | Default | Notes |
| --- | --- | --- |
| `topic` | — | Search query / subject |
| `budget_tokens` | `2000` | Clamped to `[1, 100_000]` |
| `ordering` | `"relevance+recency"` | Or `"relevance"` / `"recency"` |

Returns a markdown string, e.g.

```
## Context for 'release process' (8 memories, ~1847 tokens)

- **release/cadence**: Weekly on Thursdays at 2pm UTC
- **release/back-merge**: main → development after every prod deploy
- ...
```

## Approach

- **4-chars-per-token heuristic** instead of `tiktoken` — keeps the
  Lambda bundle slim and errs on the conservative side (the packer
  would rather leave budget on the table than overflow)
- **Greedy fit, not truncate** — a mid-stream memory too big for the
  remaining budget is skipped, not cropped. A half-quoted decision is
  worse than a missing one; the caller can rerun with a higher budget
  or different ordering
- **Redacted memories always excluded** — no `include_redacted`
  escape hatch on purpose
- **Same retrieval backbone as `search_memories`** — hybrid vector
  + keyword + recency blend; the `relevance+recency` mode matches the
  search default 1:1
- **Pure-function helpers extracted** (`estimate_tokens`,
  `pack_memories_within_budget`, `_render_packed_context`,
  `_score_for_ordering`) so the token-budget logic is unit-testable
  without auth or storage plumbing

## Test plan

- [x] `tests/unit/test_server.py::TestPackContextHelpers` — 8 tests:
  token estimator (empty / rounds-up), score-ordering dispatcher (all
  three modes + fallback), greedy pack (happy path / budget-too-small /
  mid-stream skip), render-block (empty / non-empty formatting)
- [x] `tests/unit/test_server.py::TestPackContext` — 8 integration
  tests: no-index / vector-error non-fatal, happy-path pack, budget
  overflow skips lower-ranked memory, redacted memories excluded,
  ordering modes sort candidates differently, budget clamp, invalid
  `ordering` falls back to blend
- [x] Docs: `docs-site/tools/pack-context.md` + sidebar entry + tool
  count in overview
- [x] `uv run inv pre-push` clean locally (format + lint + typecheck +
  unit + frontend)

Per §7.5, auto-merge is **not** armed at PR creation — step 7.5
arms it after the Copilot loop resolves.

https://claude.ai/code/session_01Pws345BLe4hJdH2Qqrt7qb

### CODE REVIEW

File: src/hive/server.py [hunk 1]
```diff
@@ -1375,6 +1376,209 @@ async def relate_memories(
...
+    truncate individual memories because a half-quoted decision is
+    worse than a missing one. The caller can drop the budget lower or
+    adjust ordering to surface smaller candidates.
+    """
+    packed: list[Memory] = []
+    used = 0
+    for memory, _score in memories:
+        entry_tokens = estimate_tokens(_render_memory_entry(memory))
+        if used + entry_tokens > budget_tokens:
+            continue
```
[the remark this answers is not in the document]
Comment (warlordofmars) [reply src/hive/server.py:1461]: Fixed in 53c6ca8 — the tool now reserves an over-conservative estimate for the header + trailing blank line (using sentinel-sized `000 memories, ~00000 tokens` placeholders) and subtracts it from `budget` before calling the greedy packer. The rendered output stays under the advertised budget even for small values.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/hive/server.py [hunk 2]
```diff
@@ -1375,6 +1376,209 @@ async def relate_memories(
...
+) -> tuple[list[Memory], int]:
+    """Greedily pack memories into a token budget, preserving input order.
+
+    ``memories`` is a pre-sorted list of (Memory, score) tuples; the
+    score is only carried for the header and ignored here. Returns the
+    subset that fits and the sum of their estimated tokens.
+
+    Items larger than the remaining budget are skipped — we don't
+    truncate individual memories because a half-quoted decision is
+    worse than a missing one. The caller can drop the budget lower or
```
[the remark this answers is not in the document]
Comment (warlordofmars) [reply src/hive/server.py]: Fixed in 53c6ca8 — docstring now says "raise the budget". Good catch; that was a copy-paste error from an earlier draft.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: docs-site/tools/overview.md [hunk 3]
```diff
@@ -12,6 +12,7 @@ Hive exposes six tools that your AI agent can call during a conversation. You do
 | [`list_memories`](/tools/list-memories) | List all memories with a given tag |
 | [`search_memories`](/tools/search-memories) | Search memories by semantic similarity |
 | [`summarize_context`](/tools/summarize-context) | Summarise all memories on a topic |
+| [`pack_context`](/tools/pack-context) | Token-budget-aware context pack for agents |
```
[the remark this answers is not in the document]
Comment (warlordofmars) [reply docs-site/tools/overview.md:16]: Fixed in 53c6ca8 — added `pack_context` to the `memories:read` scope grant list. Also added a natural-language hint row ("Fill my remaining context window with…").

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/unit/test_server.py [hunk 4]
```diff
@@ -2103,3 +2103,268 @@ async def test_render_prompt_roundtrip_produces_user_message(self):
...
+    def test_render_packed_context_formats_entries(self):
+        from hive.models import Memory
+        from hive.server import _render_packed_context
+
+        m = Memory(key="k", value="v", tags=[], owner_client_id="c1")
+        out = _render_packed_context("ops", [m], 3)
+        assert out.startswith("## Context for 'ops' (1 memories, ~3 tokens)")
+        assert "- **k**: v" in out
+
+
```
[the remark this answers is not in the document]
Comment (warlordofmars) [reply tests/unit/test_server.py]: Fixed in 53c6ca8 — extracted `_memory_label(count)` helper (returns `"memory"` for 1, `"memories"` otherwise) and updated the test to cover both singular and plural rendering, matching your suggestion.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: docs-site/tools/pack-context.md [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+5. The packed result is rendered as markdown with a summary header
+
+## Output shape
+
+```
+## Context for 'release process' (8 memories, ~1847 tokens)
+
+- **release/cadence**: Weekly on Thursdays at 2pm UTC
+- **release/back-merge**: main → development immediately after every prod deploy
+- **release/smoke-tests**: Playwright suite runs against dev after each deploy
```
[the remark this answers is not in the document]
Comment (warlordofmars) [reply docs-site/tools/pack-context.md:40]: Docs examples already used 2+ (e.g. "8 memories", "6 memories") so they were correct for plural. The root code-level fix in 53c6ca8 handles the singular case at runtime. No doc change needed beyond what landed for the scope table.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/hive/server.py [hunk 6]
```diff
@@ -1375,6 +1376,209 @@ async def relate_memories(
...
+
+    Separated from the tool body so the formatting is unit-testable
+    without any storage or auth plumbing.
+    """
+    if not packed:
+        return (
+            f"## Context for {topic!r} (0 memories, ~0 tokens)\n\n"
+            "_No relevant memories fit within the token budget._"
+        )
+    header = f"## Context for {topic!r} ({len(packed)} memories, ~{used_tokens} tokens)"
```
[the remark this answers is not in the document]
Comment (warlordofmars) [reply src/hive/server.py]: Fixed in 53c6ca8 — `_memory_label(count)` returns `"memory"` for 1 and `"memories"` otherwise. Applied your suggestion verbatim.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/hive/server.py [hunk 7]
```diff
@@ -1375,6 +1376,223 @@ async def relate_memories(
...
+        else ("relevance+recency")
+    )
+
+    try:
+        pairs = _vector_store().search(topic, client_id, top_k=_PACK_CONTEXT_CANDIDATE_POOL)
+    except VectorIndexNotFoundError:
+        return _tool_result(_render_packed_context(topic, [], 0), storage, client_id)
+    except Exception:
+        logger.warning("Vector search failed in pack_context (non-fatal)", exc_info=True)
+        return _tool_result(_render_packed_context(topic, [], 0), storage, client_id)
```
[the remark this answers is not in the document]
Comment (warlordofmars) [reply src/hive/server.py]: Fixed in b2e5ad9 — extracted `_render_empty_within_budget(topic, budget)` that falls back in order: full context-block → header-only → `_no context_` → empty string. The vector-error and no-index branches now call it instead of `_render_packed_context(topic, [], 0)` so a tiny `budget_tokens` can't overshoot via the error response. Four new tests pin each fallback level.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/hive/server.py [hunk 8]
```diff
@@ -1375,6 +1376,223 @@ async def relate_memories(
...
+
+def pack_memories_within_budget(
+    memories: list[tuple[Memory, float]],
+    budget_tokens: int,
+) -> tuple[list[Memory], int]:
+    """Greedily pack memories into a token budget, preserving input order.
+
+    ``memories`` is a pre-sorted list of (Memory, score) tuples; the
+    score is only carried for the header and ignored here. Returns the
+    subset that fits and the sum of their estimated tokens.
```
[the remark this answers is not in the document]
Comment (warlordofmars) [reply src/hive/server.py]: Fixed in b2e5ad9 — docstring clarifies the score is only used for upstream sort and is ignored here.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/hive/server.py [hunk 9]
```diff
@@ -1375,6 +1376,223 @@ async def relate_memories(
...
+    adjust ordering to surface smaller candidates.
+    """
+    packed: list[Memory] = []
+    used = 0
+    for memory, _score in memories:
+        entry_tokens = estimate_tokens(_render_memory_entry(memory))
+        if used + entry_tokens > budget_tokens:
+            continue
+        packed.append(memory)
+        used += entry_tokens
```
[the remark this answers is not in the document]
Comment (warlordofmars) [reply src/hive/server.py]: Fixed in b2e5ad9 — packer now adds 1 separator token per non-first entry to match what `_render_packed_context` emits between entries. The existing `test_pack_memories_greedy_fits_within_budget` was updated to assert the new total (13 + 1 + 8 = 22) so a regression on the separator accounting breaks the test.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/hive/server.py [hunk 10]
```diff
@@ -1375,6 +1376,254 @@ async def relate_memories(
...
+    # scorer first for the greedy packer to fit first.
+    scored.sort(key=lambda row: row[1], reverse=True)
+    # Reserve budget for the header + trailing blank line so the total
+    # rendered tokens stay under the advertised `budget`. Use the most
+    # expensive possible header (assumes 5-digit `used_tokens` and
+    # 3-digit count) so we never under-reserve. Floor the effective
+    # budget at 1 so degenerate callers still get an empty block.
+    header_reserve = estimate_tokens(f"## Context for {topic!r} (000 memories, ~00000 tokens)\n\n")
+    effective_budget = max(1, budget - header_reserve)
+    packed, used_tokens = pack_memories_within_budget(scored, effective_budget)
```
[the remark this answers is not in the document]
Comment (warlordofmars) [reply src/hive/server.py]: Fixed in c8c820b — when `budget <= header_reserve` the successful path now degrades through `_render_empty_within_budget(topic, budget)` instead of running the packer with an artificially floored budget. Added a belt-and-braces post-render check that also falls back to the empty ladder if `estimate_tokens(rendered) > budget` (for weird-unicode cases the heuristic under-counts). Regression test `test_tiny_budget_falls_back_to_empty_within_budget` pins the contract for budget=5.

This is iteration 3 of Copilot review per §7.5; auto-merge armed after this reply. No more Copilot rounds.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/hive/server.py [hunk 11]
```diff
@@ -1376,6 +1377,267 @@ async def relate_memories(
...
+async def pack_context(
+    topic: Annotated[str, "Topic / query to retrieve context for"],
+    budget_tokens: Annotated[
+        int,
+        "Maximum tokens in the returned context block (1–100000). Defaults to 2000.",
+    ] = _PACK_CONTEXT_DEFAULT_BUDGET,
+    ordering: Annotated[
+        str,
+        "How to rank candidates before packing: 'relevance' (pure semantic), "
+        "'recency' (pure last-accessed decay), or 'relevance+recency' (blend, default).",
```
[the remark this answers is not in the document]
Comment (warlordofmars) [reply src/hive/server.py:1532]: Fixed in `62333bd` — the `recency` annotation now reads `"last-accessed decay, falling back to updated-at decay"` so it matches what `recency_score()` actually does.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: docs-site/tools/overview.md [hunk 12]
```diff
@@ -1,6 +1,6 @@
 # MCP tools overview
 
-Hive exposes six tools that your AI agent can call during a conversation. You don't invoke these directly — your agent decides when to use them based on your instructions.
+Hive exposes seven tools that your AI agent can call during a conversation. You don't invoke these directly — your agent decides when to use them based on your instructions.
```
[the remark this answers is not in the document]
Comment (warlordofmars) [reply docs-site/tools/overview.md:3]: Fixed in `62333bd` — dropped the exact tool count. Hive exposes 15 tools now (the additional ones — `list_tags`, `relate_memories`, `memory_history`, `forget_all`, `redact_memory`, `restore_memory`, `remember_if_absent`, `ping` — have their own doc pages linked from the sidebar). Kept the summary table focused on the 7 most commonly used ones so the overview doesn't explode.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: docs-site/tools/overview.md [hunk 13]
```diff
@@ -12,14 +12,15 @@ Hive exposes six tools that your AI agent can call during a conversation. You do
...
+| [`pack_context`](/tools/pack-context) | Token-budget-aware context pack for agents |
 
 ## Scopes
 
 Each OAuth token has one or both of the following scopes:
 
 | Scope | Grants access to |
 | --- | --- |
-| `memories:read` | `recall`, `list_memories`, `search_memories`, `summarize_context` |
+| `memories:read` | `recall`, `list_memories`, `search_memories`, `summarize_context`, `pack_context` |
```
[the remark this answers is not in the document]
Comment (warlordofmars) [reply docs-site/tools/overview.md:23]: Fixed in `62333bd` — expanded both scope rows to cover the full tool surface: `memories:read` now also lists `list_tags`, `memory_history`, `relate_memories`; `memories:write` now covers `remember_if_absent`, `forget_all`, `redact_memory`, `restore_memory`.

---
_Generated by [Claude Code](https://claude.ai/code)_
